src/main/java/net/neoforged/neoforge/capabilities/BlockCapabilityCache.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities;
+
+import java.util.Objects;
+import java.util.function.BooleanSupplier;
+import net.minecraft.core.BlockPos;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.Level;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A cache for block capabilities, to be used to track capabilities at a specific position, with a specific context.
+ * <p>
+ * The cache is invalidated when the level is notified of a change via {@link Level#invalidateCapabilities(BlockPos)}.
+ * <p>
+ * Instances are automatically cleared by the garbage collector when they are no longer in use.
```

Comment:
Is this what the first `create` method that doesn't take an `isValid` makes use of to ensure that it doesn't stay around forever?

Reply:
Yeah exactly - the GC time is unreliable, so if you register an invalidation callback you must supply `isValid` - but if you don't register a callback the GC will dispose of the cache as it sees fit (it doesn't hurt to have them lying around a bit longer).

Reply:
Makes sense. Do we want to reference that bit as well from the first `create` method? (Basically saying that the GC will dispose of it as it sees fit and the other `create` method is recommended for if you know the scope of your cache usage). I am fine either way, and you can mark this after you decide whether you think it makes sense to mention this in the `create` method as well.

Reply:
The class javadoc already states that `Instances are automatically cleared by the garbage collector when they are no longer in use.`. I don't think we want to recommend the `isValid` method - I would actually recommend the simpler method because it is harder to get wrong; and in principle the GC work should be the same once the cache is not referenced anymore. Marked as resolved.